cpmpy/solvers/pumpkin.py

```diff
@@ -0,0 +1,531 @@
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations
+        cpm_cons = toplevel_list(cpm_expr)
+        supported = {"alldifferent", "cumulative", 
+                     "min", "max", "element"}
+        cpm_cons = decompose_in_tree(cpm_cons, supported=supported) # TODO: Pumpkin supports half-reified globals
+        # safening after decompose here, need to safen toplevel elements too
+        #   which come from decomposition of other global constraints...
+        cpm_cons = no_partial_functions(cpm_cons, safen_toplevel={"element"})
+        cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
+        cpm_cons = only_bv_reifies(cpm_cons)
+        cpm_cons = only_implies(cpm_cons)
+        supported_halfreif = {"or", "sum", "wsum", "sub", "mul", "div", "abs", "min", "max"}
+        cpm_cons = reify_rewrite(cpm_cons, supported=supported_halfreif) # reified element not supported yet (TODO?)
+        cpm_cons = only_numexpr_equality(cpm_cons, supported=frozenset(["sum", "wsum", "sub"]))  # supports >, <, !=
+        cpm_cons = canonical_comparison(cpm_cons) # ensure rhs is always a constant
+        return cpm_cons
+
+    def to_predicate(self, cpm_expr):
+        """
+            Convert a CPMpy expression to a Pumpkin predicate (comparison with constant)
+        """
+        from pumpkin_py import Comparator, Predicate
+
+        if isinstance(cpm_expr, _BoolVarImpl):
+            if isinstance(cpm_expr, NegBoolView):
+                lhs, comp, rhs = cpm_expr._bv, Comparator.LessThanOrEqual, 0
+            else:
+                lhs, comp, rhs = cpm_expr, Comparator.GreaterThanOrEqual, 1
+
+        elif isinstance(cpm_expr, Comparison):
+
+            lhs, rhs = cpm_expr.args
+            assert is_num(rhs), "rhs of comparison must be a constant to be a predicate"
+
+            if isinstance(lhs, Operator): # can be sum with single arg
+                if lhs.name == "sum" and len(lhs.args) == 1:
+                    lhs = lhs.args[0]
+                else:
+                    raise ValueError("Lhs of predicate should be a sum with 1 argument") # TODO: also wsum with 1 arg/mul with const?
+
+            assert isinstance(lhs, _NumVarImpl), "lhs should be variable by now"
+
+            if cpm_expr.name == "==": comp = Comparator.Equal
+            if cpm_expr.name == "<=": comp = Comparator.LessThanOrEqual
+            if cpm_expr.name == ">=": comp = Comparator.GreaterThanOrEqual
+            if cpm_expr.name == "!=": comp = Comparator.NotEqual
+            if cpm_expr.name == "<":  comp, rhs = Comparator.LessThanOrEqual, rhs - 1
+            if cpm_expr.name == ">":  comp, rhs = Comparator.GreaterThanOrEqual, rhs + 1
+        else:
+            raise ValueError(f"Cannot convert CPMpy expression {cpm_expr} to a predicate")
+
+        var = self._ivars(lhs)
+        return Predicate(var, comp, rhs)
+
+
+    def _ivars(self, cpm_var):
```

Comment:
maybe better in singular name (where you then as a special case also accept a list)? also not sure why it is semi-private with the _ivars?